### PR TITLE
Attempt to fix DuplicateAndClose_TcpServerHandler flaky test

### PR DESCRIPTION
**Pull request by rokonec**
Context:
We have noticed failure of DuplicateAndClose_TcpServerHandler test in Job https://dev.azure.com/dnceng-public/public/_build/results?buildId=696224&view=logs&j=cb76e204-ba8f-557d-eba6-9896b4f24865

Changes made (changes are isolated into commits for easier review/revert):
[Ensure RemoteInvokeHandle is disposed](https://github.com/dotnet/runtime/commit/cb3a84c71786a316926e90323481fedb97fd35a9)  - this is ensuring that Dispose of `RemoteInvokeHandle` will be called during test run, otherwise it would be called in Finalize and as it throws, it would shutdown whole xunit runner.

[Make RunCommonHostLogic synchronous](https://github.com/dotnet/runtime/commit/0d7279593e5f26689b5a96228523f707d45bc9ce) - after discussion with original author we decide reverting `RunCommonHostLogic` back to synchronous execution for simplicity reasons.

[Document RemoteInvokeHandle DisposeAsync intent.](https://github.com/dotnet/runtime/commit/babfa61e6b4f86f1461edd9e3d28cc4849cc12e2)

[Replace namedpipe IPC by Socket IPC](https://github.com/dotnet/runtime/commit/922d4be1d78db218d19b5ab9072b3ab99be91398) - as there is theory that namedpipe implementatation is root cause of this flakiness, I have decided to rule it our by replacing it by simple `Socket` based IPC. IPC is required because remote code has to start before socket is disconnect, as socket disconnection requires process ID of the other process.

Testing: 
locally by unit tests

### CODE REVIEW

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs [hunk 1]
```diff
@@ -317,44 +317,56 @@ public async Task DuplicateAndClose_TcpServerHandler(AddressFamily addressFamily
...
                 }
 
-                static async Task<int> HandlerServerCode(string ipcPipeName)
+                static async Task<int> HandlerServerCode(string ipcPortString)
                 {
-                    await using NamedPipeClientStream pipeClientStream =
-                        new NamedPipeClientStream(".", ipcPipeName, PipeDirection.In);
-                    pipeClientStream.Connect();
+                    int ipcPort = int.Parse(ipcPortString, CultureInfo.InvariantCulture);
+                    using Socket ipcClient = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
```
Comment (wfurt) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:364]: nit: I would also add `ipcClient.NoDelay = true;` (same for the other side)

**Review by wfurt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs (hunk 1 above)
Comment (rokonec) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:364]: Great hint. Done.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs [hunk 2]
```diff
@@ -317,44 +317,58 @@ public async Task DuplicateAndClose_TcpServerHandler(AddressFamily addressFamily
...
+                    // Send socketInfo to server code
+                    using var ipcStream = new NetworkStream(ipcServer, false);
+                    WriteSocketInfo(new NetworkStream(ipcServer), socketInfo);
 
                     // Send client data:
                     client.Send(TestBytes);
                 }
 
-                static async Task<int> HandlerServerCode(string ipcPipeName)
+                static async Task<int> HandlerServerCode(string ipcPortString)
```
Comment (ManickaP) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:362]: Just a question, is there a reason why we don't pass serialized `SocketInfo` directly here as a string? Completely avoiding the extra socket?
Is it change in timing of the test?

Comment (rokonec) [reply src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs:362]: Yes, I was hoping for that too, but then realized that `DuplicateAndClose` requires process ID of the other, in our case remote, process to return SocketInfo. So the other process has to start before we obtain socket info, and we hence can't pass serialized socket into to it during its construction.

**Comment by rokonec** [comment]
/ba-g After careful review I am quite sure failing tests are not related to changes of this PR

**Pull request merged**